ext/opentelemetry-ext-grpc/src/opentelemetry/ext/grpc/__init__.py

```diff
@@ -12,13 +12,61 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
 # pylint:disable=import-outside-toplevel
 # pylint:disable=import-self
+from concurrent import futures
 # pylint:disable=no-name-in-module
 # pylint:disable=relative-beyond-top-level
-
+from contextlib import contextmanager
+from wrapt import wrap_function_wrapper as _wrap
 from opentelemetry import trace
+from opentelemetry.auto_instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.ext.grpc.grpcext import intercept_channel, intercept_server
 from opentelemetry.ext.grpc.version import __version__
+from opentelemetry.instrumentation.utils import unwrap
+
+
+
+class GrpcInstrumentorServer(BaseInstrumentor):
+    def __init__(self):
+        self.server = None
+
+    def _instrument(self, **kwargs):
+        _wrap("grpc", "server", self.wrapper_fn)
+
+    def _uninstrument(self, **kwargs):
+        unwrap("grpc", "server")
+
+    def wrapper_fn(self, original_func, instance, args, kwargs):
+        self.server = original_func(*args, **kwargs)
+        self.server = intercept_server(self.server, server_interceptor())
+
+
+class GrpcInstrumentorClient(BaseInstrumentor):
+    def __init__(self):
+        self.channel = None
+
+    def _instrument(self, **kwargs):
+
+        if kwargs.get("channel_type") == "secure":
+            _wrap("grpc", "secure_channel", self.wrapper_fn)
+
+        else:
+            _wrap("grpc", "insecure_channel", self.wrapper_fn)
+
+    def _uninstrument(self, **kwargs):
+        if kwargs.get("channel_type") == "secure":
+            unwrap("grpc", "secure_channel")
+
+        else:
+            unwrap("grpc", "insecure_channel")
+
+    @contextmanager
+    def wrapper_fn(self, original_func, instance, args, kwargs):
+        with original_func(*args, **kwargs) as channel:
+            self.channel = intercept_channel(channel, client_interceptor())
+            yield self.channel
```

Comment:
I'm not too familiar with contextmanagers so can't confirm this is right (although it looks good), but definitely we shouldn't need to set `self.channel`, returning/yielding it should be good enough, if that's the behaviour of the unwrapped function

Reply:
Please see above comment as similar logic exists in the client instrumentor.